src/cache_manager.cc

```diff
@@ -0,0 +1,414 @@
+    const std::string& cache_config, std::shared_ptr<TritonCache>* cache)
+{
+  LOG_INFO << "Creating TritonCache with name: '" << name << "', libpath: '"
+           << libpath << "', cache_config: '" << cache_config << "'";
+
+  auto lcache = std::shared_ptr<TritonCache>(
+      new TritonCache(name, libpath, cache_config));
+
+  RETURN_IF_ERROR(lcache->LoadCacheLibrary());
+  RETURN_IF_ERROR(lcache->InitializeCacheImpl());
+
+  *cache = std::move(lcache);
+  return Status::Success;
+}
+
+TritonCache::TritonCache(
+    const std::string& name, const std::string& libpath,
+    const std::string& cache_config)
+    : name_(name), libpath_(libpath), cache_config_(cache_config)
+{
+  ClearHandles();
+}
+
+TritonCache::~TritonCache()
+{
+  LOG_VERBOSE(1) << "unloading cache '" << name_ << "'";
+  if (fini_fn_ != nullptr) {
+    LOG_VERBOSE(2) << "Calling TRITONCACHE_CacheDelete from: '" << libpath_
+                   << "'";
+    LOG_TRITONSERVER_ERROR(fini_fn_(cache_impl_), "failed finalizing cache");
+  } else {
+    LOG_ERROR << "cache finalize function is nullptr";
+  }
+
+  ClearHandles();
+}
+
+void
+TritonCache::ClearHandles()
+{
+  dlhandle_ = nullptr;
+  cache_impl_ = nullptr;
+  init_fn_ = nullptr;
+  fini_fn_ = nullptr;
+  lookup_fn_ = nullptr;
+  insert_fn_ = nullptr;
+}
+
+Status
+TritonCache::LoadCacheLibrary()
+{
+  LOG_VERBOSE(1) << "Loading cache library: '" << name_ << "'";
+  TritonCacheInitFn_t init_fn;
+  TritonCacheFiniFn_t fini_fn;
+  TritonCacheLookupFn_t lookup_fn;
+  TritonCacheInsertFn_t insert_fn;
+
+  // Load the library and initialize all the entrypoints
+  {
+    std::unique_ptr<SharedLibrary> slib;
+    RETURN_IF_ERROR(SharedLibrary::Acquire(&slib));
+    RETURN_IF_ERROR(slib->OpenLibraryHandle(libpath_, &dlhandle_));
+
+    // Cache initialize and finalize functions, required
+    RETURN_IF_ERROR(slib->GetEntrypoint(
+        dlhandle_, "TRITONCACHE_CacheNew", false /* optional */,
+        reinterpret_cast<void**>(&init_fn)));
+    RETURN_IF_ERROR(slib->GetEntrypoint(
+        dlhandle_, "TRITONCACHE_CacheDelete", false /* optional */,
+        reinterpret_cast<void**>(&fini_fn)));
+    RETURN_IF_ERROR(slib->GetEntrypoint(
+        dlhandle_, "TRITONCACHE_CacheLookup", false /* optional */,
+        reinterpret_cast<void**>(&lookup_fn)));
+    RETURN_IF_ERROR(slib->GetEntrypoint(
+        dlhandle_, "TRITONCACHE_CacheInsert", false /* optional */,
+        reinterpret_cast<void**>(&insert_fn)));
+  }
+
+  init_fn_ = init_fn;
+  fini_fn_ = fini_fn;
+  lookup_fn_ = lookup_fn;
+  insert_fn_ = insert_fn;
+  return Status::Success;
+}
+
+Status
+TritonCache::InitializeCacheImpl()
+{
+  if (init_fn_ == nullptr) {
+    return Status(Status::Code::NOT_FOUND, "cache init function is nullptr");
+  }
+
+  // TODO: TRITONSERVER_Message doesn't seem to do any validation on the
+  //       string to assert valid JSON. Not sure why passing
+  //       'TritonServerMessage' would be better than just passing 'const char*'
```

Comment:
@GuanLuo curious of your opinion on passing `const char* config_json` vs. `TRITONSERVER_Message* config_json` to shared library. 

I thought `TritonServerMessage` would do some JSON validation, but it doesn't, so I wasn't sure what value it added in this use case.